palette/src/relative_contrast.rs

```diff
@@ -0,0 +1,137 @@
+use white_point::WhitePoint;
+use {from_f64, FloatComponent, FromF64, IntoColor};
+
+/// A trait for calculating relative contrast between two colors.
+///
+/// W3C's Web Content Accessibility Guidelines (WCAG) 2.1 suggest a method
+/// to calculate accessible contrast ratios of text and background colors
+/// for those with low vision or color deficiencies, and for contrast of user
+/// interface graphics. The possible range of ratios is from 1:1 to 21:1.
+/// There is a Success Criterion for Contrast (Minimum) and a Success Criterion
+/// for Contrast (Enhanced), SC 1.4.3 and SC 1.4.6 respectively, which are
+/// concerned with text and images of text. SC 1.4.11 is a Success Criterion for
+/// "non-text contrast" such as user interface components and other graphics.
+/// The relative contrast is calculated by `(L1 + 0.05) / (L2 + 0.05)`, where
+/// `L1` is the luminance of the brighter color and `L2` is the luminance of the
+/// darker color both in sRGB linear space. Higher contrast ratio is generally
+/// desireable.
+///
+/// For more details of the guidelines:
+///
+/// [Success Criterion 1.4.3 Contrast (Minimum) (Level AA)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum)
+///
+/// [Success Criterion 1.4.6 Contrast (Enhanced) (Level AAA)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-enhanced)
+///
+/// [Success Criterion 1.4.11 Non-text Contrast (Level AA)](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)
+pub trait RelativeContrast<C, Wp, T>
```

Comment:
None of these should be necessary here... If the blanket implementation complains, then that's the problem we have to solve.

Reply:
If I remove them then I get:

```rust
error[E0207]: the type parameter `Wp` is not constrained by the impl trait, self type, or predicates
  --> palette/src/relative_contrast.rs:64:9
   |
64 | impl<C, Wp, T> RelativeContrast for C
   |         ^^ unconstrained type parameter

error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
  --> palette/src/relative_contrast.rs:64:13
   |
64 | impl<C, Wp, T> RelativeContrast for C
   |             ^ unconstrained type parameter

error: aborting due to 2 previous errors
```

Reply:
Oh no... That's a bummer. I can think of three options that do not involve a lot more figuring out:

1. Make `RelativeContrast` have a `<Wp, T>` parameter and remove `Scalar`. Not the best option, unless we want a many-to-many relationship between types and `RelativeContrast` implementations.
2. Give `RelativeContrast` an additional `WhitePoint` associated type and implement it for `C: IntoColor<C::WhitePoint, C::Scalar>` (I think this will work). Not fun to have a seemingly unrelated associated type. It does at least mean that there is only one implementation per scalar and white point pair.
3. Go back to implementing it for every type. Tedious but should at least work.

It may also be possible to use other traits to get the required types, but I'm not sure how to do it in a good way yet. Something like a universal `trait Color {type WhitePoint; type Component;}`, but I'm not sure if it's better or worse.

Reply:
1. If I go for the approach of 1 with removing the `Scalar` and having the `<Wp, T>` parameter, I get this which I'm not sure how to solve.
```rust
error[E0284]: type annotations needed: cannot resolve `<f32 as std::ops::Add<_>>::Output == _`
  --> palette/src/relative_contrast.rs:78:24
   |
78 |             (this.luma + from_f64(0.05)) / (other.luma + from_f64(0.05))
   |                        ^
```

2. I'm not sure how to make 2 work.
```rust
error[E0391]: cycle detected when computing the bounds for type parameter `C`
  --> palette/src/relative_contrast.rs:66:18
   |
66 |     C: IntoColor<C::WhitePoint, C::Scalar> + Copy,
   |                  ^^^^^^^^^^^^^
   |
   = note: ...which again requires computing the bounds for type parameter `C`, completing the cycle
note: cycle used when processing `relative_contrast::<impl at palette/src/relative_contrast.rs:64:1: 83:2>`
```

I can implement for every type, I was hoping not to.

Reply:
It's hard to tell without seeing the trait bounds, but the first one seem really odd. It should be the same you have, but without the `C` parameter on `RelativeContrast`. An extended variant of that is to have the trait require `IntoColor<Wp, T>` and be all default implementations. But that's kind of getting a bit weird.

I'm starting to think that implementing it for each of them is the way to go. It's tedious, and I know you already almost did before, but it's simple and it's going to be the approach that requires the least restrictions as it is today. The trait doesn't have to be aware of anything more than its output type.

What do you think?

Reply:
Sorry, I should have included gists of what I tried. I can't recreate the issue I had with `1.` but I came back with a clearer head to try `1.` again and it worked.

Including a gist of what I tried to `2.` https://gist.github.com/okaneco/2b9d8c9f1ddd2d59ea9a61404806cece

Reply:
Alright, seems like it's not possible to use the associated types like that.

To add to my reasoning, one of the downsides I see with option 1 and possibly 2 is that one would have to pick a white point to use it in trait bounds.

Reply:
I agree and favor the more permissive way as it was implemented for each type before. 

I cleaned up some of the imports to be more consistent in separating the functions, structs, and traits across the files I interacted with.